core/src/preferences/PreferencesBackend.cpp

```diff
@@ -183,87 +183,123 @@ namespace ledger {
+                readValue = [&](std::vector<uint8_t> const& value) {
+                    return value;
+                };
+            }
 
             // now we can iterate over all data, decrypt with the “old” cipher, encrypt with the
             // “new” cipher and generate two changes per entry: one that deletes it (to prevent to
             // duplicating it) and one that puts the new encoded one; once the iteration is done,
             // we submit the batch to leveldb and it applies it atomically
-            auto it = std::unique_ptr<leveldb::Iterator>(_db->NewIterator(leveldb::ReadOptions()));
             leveldb::WriteBatch batch;
+            auto it = std::unique_ptr<leveldb::Iterator>(_db->NewIterator(leveldb::ReadOptions()));
 
             for (it->SeekToFirst(); it->Valid(); it->Next()) {
-                // decrypt with the old cipher
+                // decrypt with the old cipher, if any
                 auto value = it->value().ToString();
-                auto ciphertext = std::vector<uint8_t>(std::begin(value), std::end(value));
-                auto plaindata = decrypt_preferences_change(ciphertext, *_cipher);
+                auto plaindata = std::vector<uint8_t>(value.cbegin(), value.cend());
+                plaindata = readValue(plaindata);
 
-                // the key is not encrypted
+                // the key is never encrypted
                 auto keyStr = it->key().ToString();
                 auto key = std::vector<uint8_t>(keyStr.cbegin(), keyStr.cend());
 
                 // remove the key and its associated value to prevent duplication
                 putPreferencesChange(batch, _cipher, PreferencesChange(PreferencesChangeType::DELETE_TYPE, key, {}));
 
-                // encrypt with the new cipher; in order to do that, we need a PreferencesChange
+                // encrypt with the new cipher (if any); in order to do that, we need a PreferencesChange
                 // to add with the new cipher
                 auto change = PreferencesChange(PreferencesChangeType::PUT_TYPE, key, plaindata);
                 putPreferencesChange(batch, newCipher, change);
             }
 
-            // we also need to update the salt
+            // we also need to update the salt if we are encrypting
             auto saltKey = std::vector<uint8_t>(ENCRYPTION_SALT_KEY.cbegin(), ENCRYPTION_SALT_KEY.cend());
-            auto saltChange = PreferencesChange(PreferencesChangeType::PUT_TYPE, saltKey, newSaltBytes);
-            auto noCipher = Option<AESCipher>::NONE;
-            putPreferencesChange(batch, _cipher, PreferencesChange(PreferencesChangeType::DELETE_TYPE, saltKey, {}));
-            putPreferencesChange(batch, noCipher, PreferencesChange(
-                PreferencesChangeType::PUT_TYPE, saltKey, newSaltBytes
-            ));
+            putPreferencesChange(batch, noCipher, PreferencesChange(PreferencesChangeType::DELETE_TYPE, saltKey, {}));
+
+            if (newCipher.hasValue()) {
```

Comment:
I think we always need to set this value. Image we have encrypted data + not encrypted salt. Then we decide to change password to "". So we encrypt ALL the values - we did decrypt(data) and decrypt (not encrypted salt). Now we have garbage in our salt.

Reply:
We always need to set a `newCipher`? No we don’t: when decrypting from a ciphertext DB to a plaintext one. In this case, the salt must done be written, either.

Reply:
If you change password to `""`, it will _decrypt_ all the values, not encrypt. The salt will be removed and you will get back to a plaintext db.

Reply:
I meant we always need to overwrite the salt, because during the loop we watsed it.

Reply:
I ok, saw the deletion now. My bad